### PR TITLE
Rewriting ldap authentication to be more flexible.

### PR DESCRIPTION
**Pull request by maddalab**
Summary:
More configuration parameters that allow for greater range of
ldap deployment variation for authentication.

Starting with [1] builds on it to complete the suggestion on that PR, specifically the second paragraph.

@asheepapart @byxorna @Primer42 @nsauro 

Tested with ldap authentication, these changes are backwards compatible and will not need any updates to config

[1] https://github.com/tumblr/collins/pull/100

### CODE REVIEW

**Comment by asheepapart** [comment]
@maddalab This is a great improvement. Thanks!

**Comment by byxorna** [comment]
@roymarantz 
LGTM in a cursory pass

File: app/util/security/LdapAuthenticationProvider.scala [hunk 1]
```diff
@@ -32,76 +38,116 @@ class LdapAuthenticationProvider() extends AuthenticationProvider {
...
-  protected def getSecurityPrincipal(username: String): String = {
-    "%s,%s".format(getPrincipal(username), searchbase)
-  }
-
-  // returns (&cn=*)(uniqueMember=uid=USERNAME,ou=people,dc=example,dc=org)
-  protected def groupQuery(username: String): String = {
+  protected def groupQuery(dn: String, username: String): String = {
     if (config.isRfc2307Bis)
-      "(&(cn=*)(%s=%s))".format(groupattrib, getSecurityPrincipal(username))
+      "(&(%s=*)(%s=%s))".format(groupnameattrib, groupattrib, dn)
```
Comment (roymarantz) [inline app/util/security/LdapAuthenticationProvider.scala]: dn seems like a funny attribute to use for a group lookup.  I would think basedn would be better, but I'm not sure what value it is likely to have.  Do you expect dn == 'ou=people,dc=example,dc=org' or 'ou=group,dc=example,dc=org' ?

**Comment by roymarantz** [comment]
looks pretty good, but there is lots of magic that I didn't investigate closely.

**Pull request merged**